File: wow/event/registry.go
package event
import "sync"

var (
	eventsMu sync.RWMutex
	events map[string][]Handler
)

func Register(event string, handler Handler) {
	eventsMu.Lock()
	defer eventsMu.Unlock()

	events[event] = append(events[event], handler)
}

func Trigger(event string, args ...interface{}) {
	eventsMu.RLock()
	defer eventsMu.RUnlock()

	for _, handler := range events[event] {
		handler(event, args...)
	}
}

type Handler func(event string, args ...interface{})

func init() {
	events = make(map[string][]Handler)
}

File: wow/event/registry.lua
-- Package declaration
local _event = _G["github.com/eandre/sbm/wow/event"] or {}
_G["github.com/eandre/sbm/wow/event"] = _event

local builtins = _G.lunar_go_builtins

_event.events = {}

_event.Register = function(e, handler)
    local t = _event.events[e]
    if t == nil then
        t = {}
        _event.events[e] = t
    end
    table.insert(t, handler)
end

_event.Trigger = function(e, ...)
    local es = _event.events[e]
    if es == nil then
        return
    end

    for _, handler in ipairs(es) do
        handler(e, ...)
    end
end

